nixos/modules/services/networking/wireguard.nix

```diff
@@ -182,17 +182,46 @@ let
 
   };
 
+  generatePeer = peer: [ {wireguardPeerConfig = {
+    Endpoint = peer.endpoint;
+    AllowedIPs = peer.allowedIPs;
+    PublicKey = peer.publicKey;
+    PresharedKey = mkIf (peer.presharedKey != null) peer.presharedKey;
+    PersistentKeepalive = mkIf (peer.persistentKeepalive != null) peer.persistentKeepalive;
+  };}];
+
+  generateNetDev = name: values:
+    nameValuePair "41-${name}" {
+      netdevConfig.Name = "${name}";
+      netdevConfig.Kind = "wireguard";
+
+      wireguardConfig.PrivateKey = mkIf (values.privateKey != null) values.privateKey;
+      wireguardConfig.ListenPort = mkIf (values.listenPort != null) values.listenPort;
+      wireguardConfig.FwMark = mkIf (values ? FwMark) values.FwMark;
+      wireguardPeers = concatMap generatePeer values.peers;
+    };
+
+  generateNetwork = name: values:
+    nameValuePair "41-${name}" {
+      networkConfig.Description = "WireGuard Tunnel - ${name}";
+      matchConfig.Name = "${name}";
+      address = values.ips;
+    };
+
   generateUnit = name: values:
     # exactly one way to specify the private key must be set
     assert (values.privateKey != null) != (values.privateKeyFile != null);
-    let privKey = if values.privateKeyFile != null then values.privateKeyFile else pkgs.writeText "wg-key" values.privateKey;
-    in
     nameValuePair "wireguard-${name}"
       {
         description = "WireGuard Tunnel - ${name}";
-        after = [ "network.target" ];
+        after = [ "sys-subsystem-net-devices-${name}.device" "systemd-networkd-online" ];
```

Comment:
The target is spelled wrong and I don't see why `systemd-networkd-online.target` is required.

Reply:
Also use `utils.escapeSystemdPath` on name in the unit definition as systemd will escape e.g. hyphens with \x2d